homeassistant/components/matter/adapter.py

```diff
@@ -52,13 +52,62 @@ def register_platform_handler(
 
     async def setup_nodes(self) -> None:
         """Set up all existing nodes and subscribe to new nodes."""
-        for node in await self.matter_client.get_nodes():
+        for node in self.matter_client.get_nodes():
             self._setup_node(node)
 
         def node_added_callback(event: EventType, node: MatterNode) -> None:
             """Handle node added event."""
             self._setup_node(node)
 
+        def endpoint_added_callback(event: EventType, data: dict[str, int]) -> None:
+            """Handle endpoint added event."""
+            node = self.matter_client.get_node(data["node_id"])
+            self._setup_endpoint(node.endpoints[data["endpoint_id"]])
+
+        def endpoint_removed_callback(event: EventType, data: dict[str, int]) -> None:
+            """Handle endpoint removed event."""
+            server_info = cast(ServerInfoMessage, self.matter_client.server_info)
+            try:
+                node = self.matter_client.get_node(data["node_id"])
+            except KeyError:
+                return  # race condition
+            device_registry = dr.async_get(self.hass)
+            endpoint = node.endpoints.get(data["endpoint_id"])
+            if not endpoint:
+                return  # race condition
+            node_device_id = get_device_id(
+                server_info,
+                node.endpoints[data["endpoint_id"]],
+            )
+            identifier = (DOMAIN, f"{ID_TYPE_DEVICE_ID}_{node_device_id}")
+            if device := device_registry.async_get_device({identifier}):
+                device_registry.async_remove_device(device.id)
```

Comment:
We normally just remove the config entry id from device entry, and let the registry remove the device if no other config entries are attached.